pydantic_ai_slim/pydantic_ai/agent/__init__.py

```diff
@@ -1333,6 +1344,37 @@ def _normalize_instructions(
             return [instructions]
         return list(instructions)
 
+    def _apply_force_first_request(self, model_settings: ModelSettings | None) -> ModelSettings | None:
+        """Apply force_first_request tool flags to model_settings.
+
+        If any tools have force_first_request=True and no callable tool_choice is set,
+        generate a callable that forces those tools on the first request.
+        """
+        force_first_tools = [tool.name for tool in self._function_toolset.tools.values() if tool.force_first_request]
+
+        if not force_first_tools:
+            return model_settings
+
+        user_tool_choice = model_settings.get('tool_choice') if model_settings else None
+
+        if callable(user_tool_choice):
+            warnings.warn(
+                'force_first_request on tools is ignored when a callable tool_choice is set in model_settings',
+                UserWarning,
+                stacklevel=4,
+            )
+            return model_settings
+
+        # Generate callable that handles first request
+        def generated_tool_choice(ctx: RunContext[Any]) -> ToolChoiceValue:
+            if ctx.run_step == 1:
```

Comment:
See https://github.com/pydantic/pydantic-ai/issues/2799#issuecomment-3453795860:

> I initially thought that `prepare_model_settings` (or something like that) + checking `ctx.run_step` would likely be sufficient to let people do that manually, but if a tool calls fails, we have to keep telling the model to use it until it calls it successfully, so tracking `run_step` isn't enough.

This is the naive approach that's insufficient because the model may have _unsuccessfully_ called the tool in the first run step, and we should keep requiring it until it calls it successfully.

The commend just before that illustrates the work someone went through to work around this: https://github.com/pydantic/pydantic-ai/issues/2799#issuecomment-3452886142